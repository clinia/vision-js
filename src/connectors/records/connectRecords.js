import {
  checkRendering,
  createDocumentationMessageGenerator,
  addAbsolutePosition,
  addQueryID,
  noop,
} from '../../lib/utils';

const withUsage = createDocumentationMessageGenerator({
  name: 'records',
  connector: true,
});

/**
 * @typedef {Object} RecordsRenderingOptions
 * @property {Object[]} records The matched records from Clinia API.
 * @property {Object} results The complete results response from Clinia API.
 * @property {Object} widgetParams All original widget options forwarded to the `renderFn`.
 */

/**
 * @typedef {Object} CustomRecordsWidgetOptions
 * @property {function(Object[]):Object[]} [transformItems] Function to transform the items passed to the templates.
 */

/**
 * **Records** connector provides the logic to create custom widgets that will render the results retrieved from Clinia.
 * @type {Connector}
 * @param {function(RecordsRenderingOptions, boolean)} renderFn Rendering function for the custom **Records** widget.
 * @param {function} unmountFn Unmount function called when the widget is disposed.
 * @return {function(CustomRecordsWidgetOptions)} Re-usable widget factory for a custom **Records** widget.
 * @example
 * // custom `renderFn` to render the custom Records widget
 * function renderFn(RecordsRenderingOptions) {
 *   RecordsRenderingOptions.widgetParams.containerNode.html(
 *     RecordsRenderingOptions.records.map(function(record) {
 *       return '<div>' + records._highlightResult.name.value + '</div>';
 *     })
 *   );
 * }
 *
 * // connect `renderFn` to Records logic
 * var customRecords = vision.connectors.connectRecords(renderFn);
 *
 * // mount widget on the page
 * search.addWidgets([
 *   customRecords({
 *     containerNode: $('#custom-records-container'),
 *   })
 * ]);
 */
export default function connectRecords(renderFn, unmountFn = noop) {
  checkRendering(renderFn, withUsage());

  return (widgetParams = {}) => {
    const { transformItems = items => items } = widgetParams;

    return {
      $$type: 'cvi.records',

      init({ visionInstance }) {
        renderFn(
          {
            records: [],
            results: undefined,
            visionInstance,
            widgetParams,
          },
          true
        );
      },

      render({ results, visionInstance }) {
        const initialEscaped = results.records.__escaped;

        results.records = addAbsolutePosition(
          results.records,
          results.page,
          results.perPage
        );

        results.records = addQueryID(results.records, results.queryId);

        results.records = transformItems(results.records);

        // Make sure the escaped tag stays, even after mapping over the records.
        // This prevents the records from being double-escaped if there are multiple
        // records widgets mounted on the page.
        results.records.__escaped = initialEscaped;

        renderFn(
          {
            records: results.records,
            results,
            visionInstance,
            widgetParams,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        return state;
      },

      getWidgetSearchParameters(state) {
        return state;
      },
    };
  };
}
