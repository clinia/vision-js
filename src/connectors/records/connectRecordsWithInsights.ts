import { withInsights } from '../../lib/insights';
import connectRecords from './connectRecords';

const connectRecordsWithInsights = withInsights(connectRecords);

export default connectRecordsWithInsights;
