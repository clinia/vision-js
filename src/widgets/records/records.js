/** @jsx h */

import { h, render } from 'preact';
import cx from 'classnames';
import connectRecords from '../../connectors/records/connectRecords';
import Records from '../../components/Records/Records';
import defaultTemplates from './defaultTemplates';
import {
  prepareTemplateProps,
  getContainerNode,
  createDocumentationMessageGenerator,
} from '../../lib/utils';
import { component } from '../../lib/suit';
import { withInsights, withInsightsListener } from '../../lib/insights';

const withUsage = createDocumentationMessageGenerator({ name: 'records' });
const suit = component('Records');
const RecordsWithInsightsListener = withInsightsListener(Records);

const renderer = ({ renderState, cssClasses, containerNode, templates }) => (
  { records: receivedRecords, results, visionInstance, insights },
  isFirstRendering
) => {
  if (isFirstRendering) {
    renderState.templateProps = prepareTemplateProps({
      defaultTemplates,
      templatesConfig: visionInstance.templatesConfig,
      templates,
    });
    return;
  }

  render(
    <RecordsWithInsightsListener
      cssClasses={cssClasses}
      records={receivedRecords}
      results={results}
      templateProps={renderState.templateProps}
      insights={insights}
    />,
    containerNode
  );
};

/**
 * @typedef {Object} RecordsCSSClasses
 * @property {string|string[]} [root] CSS class to add to the wrapping element.
 * @property {string|string[]} [emptyRoot] CSS class to add to the wrapping element when no results.
 * @property {string|string[]} [list] CSS class to add to the list of results.
 * @property {string|string[]} [item] CSS class to add to each result.
 */

/**
 * @typedef {Object} RecordsTemplates
 * @property {string|function(object):string} [empty=''] Template to use when there are no results.
 * @property {string|function(object):string} [item=''] Template to use for each result. This template will receive an object containing a single record. The record will have a new property `__recordIndex` for the position of the record in the list of displayed records.
 */

/**
 * @typedef {Object} RecordsWidgetOptions
 * @property {string|HTMLElement} container CSS Selector or HTMLElement to insert the widget.
 * @property {RecordsTemplates} [templates] Templates to use for the widget.
 * @property {RecordsCSSClasses} [cssClasses] CSS classes to add.
 * @property {boolean} [escapeHTML = true] Escape HTML entities from records string values.
 * @property {function(object[]):object[]} [transformItems] Function to transform the items passed to the templates.
 */

/**
 * Display the list of results (records) from the current search.
 *
 * This is a traditional display of the records. It has to be implemented
 * together with a pagination widget, to let the user browse the results
 * beyond the first page.
 * @type {WidgetFactory}
 * @devNovel Records
 * @category basic
 * @param {RecordsWidgetOptions} $0 Options of the Records widget.
 * @return {Widget} A new instance of Records widget.
 * @example
 * search.addWidgets([
 *   vision.widgets.records({
 *     container: '#records-container',
 *     templates: {
 *       empty: 'No results',
 *       item: '<strong>Record {{id}}</strong>: {{{_highlightResult.name.value}}}'
 *     },
 *     transformItems: items => items.map(item => item),
 *   })
 * ]);
 */
export default function records({
  container,
  escapeHTML,
  transformItems,
  templates = defaultTemplates,
  cssClasses: userCssClasses = {},
}) {
  if (!container) {
    throw new Error(withUsage('The `container` option is required.'));
  }

  const containerNode = getContainerNode(container);
  const cssClasses = {
    root: cx(suit(), userCssClasses.root),
    emptyRoot: cx(suit({ modifierName: 'empty' }), userCssClasses.emptyRoot),
    list: cx(suit({ descendantName: 'list' }), userCssClasses.list),
    item: cx(suit({ descendantName: 'item' }), userCssClasses.item),
  };

  const specializedRenderer = renderer({
    containerNode,
    cssClasses,
    renderState: {},
    templates,
  });

  const makeRecords = withInsights(connectRecords)(specializedRenderer, () =>
    render(null, containerNode)
  );

  return makeRecords({ escapeHTML, transformItems });
}
