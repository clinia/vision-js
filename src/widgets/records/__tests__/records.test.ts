import { render } from 'preact';
import defaultTemplates from '../defaultTemplates';
import records from '../records';

jest.mock('preact', () => {
  const module = require.requireActual('preact');

  module.render = jest.fn();

  return module;
});

describe('Usage', () => {
  it('throws without container', () => {
    expect(() => {
      // @ts-ignore
      records({ container: undefined });
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('records()', () => {
  let container;
  let templateProps;
  let widget;
  let results;

  beforeEach(() => {
    render.mockClear();

    container = document.createElement('div');
    templateProps = {
      templatesConfig: undefined,
      templates: defaultTemplates,
      useCustomCompileOptions: { item: false, empty: false },
    };
    widget = records({ container, cssClasses: { root: ['root', 'cx'] } });
    widget.init({ visionInstance: { templateProps } });
    results = {
      records: [{ first: 'hit', second: 'hit' }],
      perPage: 4,
      page: 2,
    };
  });

  it('calls twice render(<Records props />, container)', () => {
    widget.render({ results });
    widget.render({ results });

    const [firstRender, secondRender] = render.mock.calls;

    expect(render).toHaveBeenCalledTimes(2);
    expect(firstRender[0].props).toMatchSnapshot();
    expect(firstRender[1]).toEqual(container);
    expect(secondRender[0].props).toMatchSnapshot();
    expect(secondRender[1]).toEqual(container);
  });

  it('renders transformed items', () => {
    widget = records({
      container,
      transformItems: items =>
        items.map(item => ({ ...item, transformed: true })),
    });

    widget.init({ visionInstance: {} });
    widget.render({ results });

    const [firstRender] = render.mock.calls;

    expect(firstRender[0].props).toMatchSnapshot();
  });

  it('should add __position key with absolute position', () => {
    results = { ...results, page: 4, perPage: 10 };
    const state = { page: results.page };

    widget.render({ results, state });

    expect(results.records[0].__position).toEqual(41);
  });
});
