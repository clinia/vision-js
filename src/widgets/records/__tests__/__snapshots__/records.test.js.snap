// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Usage throws without container 1`] = `
"The \`container\` option is required.

See documentation: https://www.clinia.com/doc/api-reference/widgets/records/js/"
`;

exports[`records() calls twice render(<Records props />, container) 1`] = `
Object {
  "cssClasses": Object {
    "emptyRoot": "cvi-Records--empty",
    "item": "cvi-Records-item",
    "list": "cvi-Records-list",
    "root": "cvi-Records root cx",
  },
  "insights": undefined,
  "records": Array [
    Object {
      "__position": 9,
      "first": "record",
      "second": "record",
    },
  ],
  "results": Object {
    "page": 2,
    "perPage": 4,
    "records": Array [
      Object {
        "__position": 9,
        "first": "record",
        "second": "record",
      },
    ],
  },
  "templateProps": Object {
    "templates": Object {
      "empty": "No results",
      "item": [Function],
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "empty": false,
      "item": false,
    },
  },
}
`;

exports[`records() calls twice render(<Records props />, container) 2`] = `
Object {
  "cssClasses": Object {
    "emptyRoot": "cvi-Records--empty",
    "item": "cvi-Records-item",
    "list": "cvi-Records-list",
    "root": "cvi-Records root cx",
  },
  "insights": undefined,
  "records": Array [
    Object {
      "__position": 9,
      "first": "record",
      "second": "record",
    },
  ],
  "results": Object {
    "page": 2,
    "perPage": 4,
    "records": Array [
      Object {
        "__position": 9,
        "first": "record",
        "second": "record",
      },
    ],
  },
  "templateProps": Object {
    "templates": Object {
      "empty": "No results",
      "item": [Function],
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "empty": false,
      "item": false,
    },
  },
}
`;

exports[`records() renders transformed items 1`] = `
Object {
  "cssClasses": Object {
    "emptyRoot": "cvi-Records--empty",
    "item": "cvi-Records-item",
    "list": "cvi-Records-list",
    "root": "cvi-Records",
  },
  "insights": undefined,
  "records": Array [
    Object {
      "__position": 9,
      "first": "record",
      "second": "record",
      "transformed": true,
    },
  ],
  "results": Object {
    "page": 2,
    "perPage": 4,
    "records": Array [
      Object {
        "__position": 9,
        "first": "record",
        "second": "record",
        "transformed": true,
      },
    ],
  },
  "templateProps": Object {
    "templates": Object {
      "empty": "No results",
      "item": [Function],
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "empty": false,
      "item": false,
    },
  },
}
`;
